applybn/feature_selection/bn_nmi_feature_selector.py

```diff
@@ -0,0 +1,166 @@
+    self.bin_labels = [i for i in range(self.bin_count)]
+    self.verboose = verboose
+    self.entropy_map = dict()
+
+
+  def _discreticise(self, x, y=None):
+    '''
+    Prepare data to feed into pyitlib drv.entropy
+    Calculate happenings of each event in descreticised feature
+    '''
+    arr_x = np.array(x, dtype=int)
+    
+    if y is None:
+      # if only one given, return bincount integers from 0 to max(x)
+      return np.bincount(arr_x, minlength=max(1, max(x) - min(x) + 1)).tolist()
+    else:
+      arr_y = np.array(y, dtype=int)
+      x_min, x_max = arr_x.min(), arr_x.max()
+      y_min, y_max = arr_y.min(), arr_y.max()
+      
+      hist, _, _ = np.histogram2d(
+        arr_x, arr_y,
+        bins=[x_max - x_min + 1, y_max - y_min + 1],
+        range=[[x_min, x_max + 1], [y_min, y_max + 1]]
+      )
+      return hist.flatten().astype(int).tolist()
+
+
+  def _normalized_mutual_information(self, a, b):
+    '''
+    Calculate Normalized Mutual Information (NMI) for pair of features
+
+    Args: 
+        a, b - pair of pd.Series to calculate NMI.
+    '''
+    if (a.name, None) not in self.entropy_map:
+      self.entropy_map[(a.name, None)] = float(drv.entropy(self._discreticise(a)))
+    if (b.name, None) not in self.entropy_map:
+      self.entropy_map[(b.name, None)] = float(drv.entropy(self._discreticise(b)))
+    if (a.name, b.name) not in self.entropy_map:
+      self.entropy_map[(a.name, b.name)] = float(drv.entropy_joint(self._discreticise(a,b)))
+
+    Ha = self.entropy_map[(a.name, None)]
+    Hb = self.entropy_map[(b.name, None)]
+    Hab = self.entropy_map[(a.name, b.name)]
+    return (Ha + Hb - Hab) / (2 * max(Ha, Hb))
+
+  def _prepare_dataset(self, df: pd.DataFrame):
+    '''
+    Prepare the dataset, discretize the fields, remove rows
+    Return a ready-to-use dataset
+    
+    Args: 
+        df - data frame to be cleaned.
+    '''
+    clean_df = pd.DataFrame()
+    for col in df:
+      if pd.api.types.is_integer_dtype(df[col].dtype):
+        # integers can be discreticised this way, without binning
+        df_map = { l:i for i,l in enumerate(df[col].unique())}
+        clean_df[col] = df[col].map(df_map)
+      elif pd.api.types.is_numeric_dtype(df[col].dtype):
+        # discretisize to bin_count bins
+        clean_df[col] = pd.cut(df[col], bins=self.bin_count, labels=self.bin_labels, duplicates='drop')
+        clean_df[col] = clean_df[col].cat.add_categories(self.bin_count + 1).fillna(self.bin_count + 1)
+      else:
+        # pandas cut does not work with string features, this is done manually
+        # it is assumed that the user has removed the ID-like features, otherwise there will be a memory error later
+        df_map = { l:i for i,l in enumerate(df[col].unique())}
+        clean_df[col] = df[col].map(df_map)
+    return clean_df
+
+  def fit(self, X: pd.DataFrame, y: pd.Series = None):    
+    if y is None:
+      return self
+
+    df = X
+    df[y.name] = y
```

Comment:
Since X is a dataframe, it is also mutable. When a user passes his data into this function, the original data will be modified later on. Because variables data, X, and df will refer to the same DataFrame object in memory. This is an undesirable behavior in most cases.
TL;DR use pandas `.copy()` method.